service/worker/batcher/activities.go

```diff
@@ -308,3 +332,97 @@ func isDone(ctx context.Context) bool {
 		return false
 	}
 }
+
+func getResetEventIDByType(ctx context.Context,
+	resetType enumspb.ResetType,
+	namespaceStr string,
+	workflowExecution *commonpb.WorkflowExecution,
+	frontendClient workflowservice.WorkflowServiceClient,
+	logger log.Logger) (int64, error) {
+	switch resetType {
+	case enumspb.RESET_TYPE_FIRST_WORKFLOW_TASK:
+		return getFirstWorkflowTaskEventID(ctx, namespaceStr, workflowExecution, frontendClient, logger)
+	case enumspb.RESET_TYPE_LAST_WORKFLOW_TASK:
+		return getLastWorkflowTaskEventID(ctx, namespaceStr, workflowExecution, frontendClient, logger)
+	default:
+		errorMsg := fmt.Sprintf("provided reset type (%v) is not supported.", resetType)
+		return 0, serviceerror.NewInvalidArgument(errorMsg)
+	}
+}
+
+func getLastWorkflowTaskEventID(ctx context.Context,
+	namespaceStr string,
+	workflowExecution *commonpb.WorkflowExecution,
+	frontendClient workflowservice.WorkflowServiceClient,
+	logger log.Logger) (workflowTaskEventID int64, err error) {
+	req := &workflowservice.GetWorkflowExecutionHistoryReverseRequest{
+		Namespace:       namespaceStr,
+		Execution:       workflowExecution,
+		MaximumPageSize: 1000,
+		NextPageToken:   nil,
+	}
+	for {
+		resp, err := frontendClient.GetWorkflowExecutionHistoryReverse(ctx, req)
+		if err != nil {
+			logger.Error("failed to run GetWorkflowExecutionHistoryReverse")
+			return 0, errors.New("failed to get workflow execution history")
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			switch e.GetEventType() {
+			case enumspb.EVENT_TYPE_WORKFLOW_TASK_COMPLETED:
+				workflowTaskEventID = e.GetWorkflowTaskCompletedEventAttributes().GetScheduledEventId() + 1
```

Comment:
As I commented in the other thread, I believe resetting to workflow completed event is always better than resetting to the corresponding scheduled event since either way new workflow history will contain a workflow task failed event and resetting to completed can reuse more history events.

Reply:
thanks for pointing that out, I missed your other comment, and I agree that reusing more of the history events is more efficient. I update the code to reflect that.